Source/BoundaryConditions/WarpX_PEC.H

```diff
@@ -295,6 +309,345 @@ using namespace amrex;
+                               or a guard cell
+     * \param[in] is_nodal     staggering of the field data being updated.
+     * \param[in] fbndry_lo    Field boundary type at the lower boundaries
+     * \param[in] fbndry_hi    Field boundary type at the upper boundaries
+     * \param[in] pbndry_lo    Particle boundary type at the lower boundaries
+     * \param[in] pbndry_hi    Particle boundary type at the upper boundaries
+     */
+    AMREX_GPU_DEVICE AMREX_FORCE_INLINE
+    void SetRhofieldFromPEC (const amrex::IntVect & dom_lo,
+                             const amrex::IntVect & dom_hi,
+                             const amrex::IntVect & nguard,
+                             const amrex::IntVect & ijk_vec,
+                             const int n,
+                             amrex::Array4<amrex::Real> const& rho,
+                             const amrex::IntVect & is_nodal,
+                             amrex::GpuArray<int, 3> const& fbndry_lo,
+                             amrex::GpuArray<int, 3> const& fbndry_hi,
+                             amrex::GpuArray<ParticleBoundaryType, 3> const& pbndry_lo,
+                             amrex::GpuArray<ParticleBoundaryType, 3> const& pbndry_hi )
+    {
+        amrex::IntVect ijk_mirror;
+        // Loop over all dimensions
+        for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+            ijk_mirror = ijk_vec;
+            // Loop over sides, iside = 0 (lo), iside = 1 (hi)
+            for (int iside = 0; iside < 2; ++iside) {
+                // check if the boundary in this direction is a PEC, otherwise
+                // continue to the next direction
+                if (!( (iside == 0 )
+                    ? is_boundary_PEC(fbndry_lo, idim)
+                    : is_boundary_PEC(fbndry_hi, idim) )) continue;
+
+                // Mirror guard cell outside the domain by "ig" number of cells,
+                // in direction, idim.
+                // Note: For nodal fields ig=0 is on the boundary and will be
+                // handled in `SetRhoFieldOnPEC` below. For cell-centered
+                // fields ig=0 is the cell center next to the PEC (outside the
+                // domain) and so it mirrors the first valid entry in the domain
+                // and should be reflected back into the domain.
+                const int ig = ( (iside == 0)
+                                ? (ijk_vec[idim] - dom_lo[idim])
+                                : (dom_hi[idim] + is_nodal[idim] - ijk_vec[idim]) );
+
+                if (ig >= is_nodal[idim] && ig <= nguard[idim]) {
+
+                    // Mirror location outside the domain by "ig" number of cells
+                    // across PEC boundary in direction, idim, and side, iside
+                    ijk_mirror[idim] = ( (iside == 0)
+                                     ? (dom_lo[idim] - (1 - is_nodal[idim]) - ig)
+                                     : (dom_hi[idim] + is_nodal[idim] + ig));
```

Comment:
would the same suggestions from above apply here as well?

Reply:
No, you are right, this should be `dom_hi[idim] + 1 + ig`. It doesn't show up as a problem since rho is always nodal. But for the sake of accuracy I'll update this to be consistent with the other cases.